pallets/corporate-actions/src/distribution.rs

```diff
@@ -0,0 +1,488 @@
+                // Progress won't happen if an error occurs for all DIDs.
+                .filter(|did| Self::transfer_share(*did, cp_id, supply, ca_id, &ca, &mut dist).is_err())
+                .inspect(|did| Self::deposit_event(Event::<T>::SharePushFailed(caa_ish, ca_id, *did)))
+                .count() as u32;
+
+            // Commit `dist` changes.
+            if count - count_fail > 0 {
+                <Distributions<T>>::insert(ca_id, dist);
+            }
+
+            // Emit some stats re. how pushing went overall.
+            Self::deposit_event(Event::<T>::SharesPushed(caa_ish, ca_id, max, count, count_fail));
+        }
+
+        /// Assuming a distribution has expired,
+        /// unlock the remaining amount in the distributing-from portfolio.
+        ///
+        /// ## Arguments
+        /// - `origin` which must be the creator of the capital distribution tied to `ca_id`.
+        /// - `ca_id` identifies the CA with a capital distributions to reclaim for.
+        ///
+        /// # Errors
+        /// - `NoSuchDistribution` if there's no capital distribution for `ca_id`.
+        /// - `NotDistributionCreator` if `origin` is not the original creator of the distribution.
+        /// - `AlreadyReclaimed` if this function has already been called successfully.
+        /// - `NotExpired` if `now < expiry`.
+        #[weight = 900_000_000]
+        pub fn reclaim(origin, ca_id: CAId) {
+            // Ensure DID is the dist creator, they haven't reclaimed, and that expiry has passed.
+            let did = <Identity<T>>::ensure_perms(origin)?;
+            let dist = Self::ensure_distribution_exists(ca_id)?;
+            ensure!(did == dist.from.did, Error::<T>::NotDistributionCreator);
+            let did = did.for_event();
+            ensure!(!dist.reclaimed, Error::<T>::AlreadyReclaimed);
+            ensure!(expired(dist.expires_at, <Checkpoint<T>>::now_unix()), Error::<T>::NotExpired);
+
+            // Unlock `remaining` of `currency` from DID's portfolio.
+            // This cannot fail, as we've already locked the requisite amount prior.
+            <Portfolio<T>>::unlock_tokens(&dist.from, &dist.currency, &dist.remaining).unwrap();
+
+            // Zero `remaining` + note that we've reclaimed.
+            <Distributions<T>>::insert(ca_id, Distribution { reclaimed: true, remaining: 0.into(), ..dist });
+
+            // Emit event.
+            Self::deposit_event(Event::<T>::Reclaimed(did, ca_id, dist.remaining));
+        }
+    }
+}
+
+decl_event! {
+    pub enum Event<T>
+    where
+        Balance = <T as CommonTrait>::Balance,
+    {
+        /// A capital distribution, with details included,
+        /// was created by the DID (the CAA) for the CA specified by the `CAId`.
+        ///
+        /// (CAA of CAId's ticker, CA's ID, distribution details)
+        Created(EventOnly<IdentityId>, CAId, Distribution<Balance>),
```

Comment:
I like how `EventOnly` is implemented however I rather meant for `EventOnly<IdentityId>` to be the type of function arguments in functions that only use those arguments for event construction either by calling `risky_to_inner` or by direct value passing, in which case the type in the event could indeed be `EventOnly<IdentityId>`.

If there are no such functions here then the type of identity ID in the events themselves should remain `IdentityId` or can change to `Option<IdentityId>` in cases where the current identity is taken from `Context` and the function may be called with undefined `Context`.

Reply:
Sure, I think it would be great as a function parameter too, but in the case of capital distributions, there were no applicable instances.

Even so, I think its better to uniformly use `EventOnly<IdentityId>` in events, even if not passed as an argument, as the re-binding strategy that I've used here, e.g., `let did = did.for_event();` ASAP provides decent protection. The strategy of doing `let did = <Identity<T>>::ensure_perms(origin)?.for_event();` provides even better protection, as it immediately converts it to `EventOnly<_>`.